packages/@aws-cdk/aws-ec2/lib/launch-template.ts

```diff
@@ -591,9 +591,14 @@ export class LaunchTemplate extends Resource implements ILaunchTemplate, iam.IGr
       },
     });
 
-    if (props.userData) {
-      this.userData = props.userData;
+    const imageConfig: MachineImageConfig | undefined = props.machineImage?.getImage(this);
+    if (imageConfig) {
+      this.osType = imageConfig.osType;
+      this.imageId = imageConfig.imageId;
     }
+
+    // priority: prop.userData -> userData from machineImage -> undefined
+    this.userData = props.userData ?? imageConfig?.userData;
```

Comment:
Why can't we do

```suggestion
    if (props.userData || imageConfig?.userData) {
      this.userData = props.userData ?? imageConfig?.userData;
    }
```

and avoid the change to the templates? I understand that those won't require replacement or interruption, but we try to avoid changes to your diff caused only by upgrading. We'd have to put this behind a feature flag.